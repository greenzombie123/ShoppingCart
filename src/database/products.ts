export type ProductCategory = "Jewelry" | "Men's Clothing" | "Women's Clothing" | "Electronics"

export type Product = {
	id:number,
	name:string,
	price:number,
	rating:number,
	stars:1|2|3|4|5,
	likes:number,
	styles:Style[]
	category:ProductCategory
}

export type Style = {
	description:string,
	picture:URL,
	isCurrentStyle:boolean, 
}

export type CartItem = {
	product:Product,
	id:number,
	quantity:number,
	style:string
}

export type Cart = CartItem[]