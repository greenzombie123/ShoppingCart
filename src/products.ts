// import fs from "fs";

export type ProductCategory =
  | "Jewelry"
  | "Men's Clothing"
  | "Women's Clothing"
  | "Electronics";

export type Product = {
  id: number | string;
  name: string;
  price: number;
  ratings: number;
  stars: 1 | 2 | 3 | 4 | 5;
  likes: number;
  styles: Style[];
  category: ProductCategory;
};

export type Style = {
  description: string;
  picture: string;
  isCurrentStyle: boolean;
};

export type CartItem = {
  name:string,
  price: number;
  id: number;
  quantity: number;
  style: string | undefined,
  picture:string
};

export type Cart = CartItem[];

// const thing: Product[] = [
//   {
//     id: 1,
//     name: "polo shirt",
//     price: 2314,
//     ratings: 334,
//     stars: 5,
//     likes: 380,
//     category: "Men's Clothing",
//     styles: [
//       {
//         description: "black",
//         picture: "/images/poloshirtblack.webp",
//         isCurrentStyle: true,
//       },
//       {
//         description: "blue",
//         picture: "/images/bluepoloshirt.jpg",
//         isCurrentStyle: false,
//       },
//     ],
//   },
//   {
//     id: 1,
//     name: "shirt",
//     price: 1299,
//     ratings: 234,
//     stars: 4,
//     likes: 1000,
//     category: "Men's Clothing",
//     styles: [
//       {
//         description: "white",
//         picture: "src/assets/bluepoloshirt.jpg",
//         isCurrentStyle: true,
//       },
//     ],
//   },
// ];

// const doShit = async () => {
//   const g = await fetch(`http://localhost:3000/test`, {
//     method: "PATCH",
//     headers: { "Content-Type": "application/json" },
//     body:JSON.stringify({test:[]})
//   }).catch(e=>console.log(e))



//   console.log(g)

// };

// doShit()
