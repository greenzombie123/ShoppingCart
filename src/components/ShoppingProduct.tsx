import React, { useState } from "react";
import { Product} from "../products";
import style from "./ShoppingProduct.module.css";
import { useLocation } from "react-router-dom";

export const Picture = ({product}: {product:Product}) => {
  const currentStyle = product.styles.filter(productStyle=>productStyle.isCurrentStyle)[0]

  return (
    <div className={style.picture}>
      <div className={style.imgContainer}>
        <img src={currentStyle.picture} alt={product.name + " " + currentStyle.description} />
      </div>

      <div className={style.colorTabs}>
        Style:
        <button className={style.colorTab}>Red</button>
      </div>
    </div>
  );
};

export const ColorTabs = () => {
  return (
    <div className={style.colorTabs}>
      Style:
      <button className={style.colorTab}>Red</button>
    </div>
  );
};

const ShoppingProduct = () => {
  const location = useLocation()
  const state = location.state as Product
  const [product, setProduct] = useState<Product>(state)

  return (
    <div className={style.shoppingProduct}>
      <Picture product={product}/>
      {/* <div className={style.picture}>
        <div className={style.imgContainer}>
          <img src="" alt="" />
        </div>
        <div className={style.colorTabs}>
          Style:
          <button className={style.colorTab}>Red</button>
        </div>
      </div> */}
      <div className={style.productdetails}>
        <p>Leather Bag</p>
        <div className={style.stars}>
          <div className={style.stars_container}>
            <div className={style.star + style.on}></div>
            <div className={style.star + style.on}></div>
            <div className={style.star + style.on}></div>
            <div className={style.star + style.on}></div>
            <div className={style.star + style.on}></div>
          </div>
          <div className={style.ratings}>(13)</div>
        </div>
        <p>$200</p>
        <p className={style.style}>Style: Red</p>
      </div>
      <div className={style.productToCart}>
        <p className={style.quantity}>Quantity</p>
        <div className={style.quantityCounter}>
          <button className={style.increaseButton}>&#43;</button>
          <p className={style.counter}>1</p>
          <button className={style.decreaseButton}>&#8722;</button>
        </div>
        <button className={style.addButton}>Add to Cart</button>
      </div>
    </div>
  );
};

export default ShoppingProduct;
