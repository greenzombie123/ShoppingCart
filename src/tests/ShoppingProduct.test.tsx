import { afterAll, beforeAll, beforeEach, describe, expect, it, Mock, vi } from "vitest";
import { Cart, Product } from "../products";
import {
  BrowserRouter,
  RouteObject,
} from "react-router-dom";
import ShoppingProduct from "../components/ShoppingProduct";
import { render, screen, waitFor } from "@testing-library/react";
import React from "react";

const product: Product = {
  id: 1,
  name: "polo shirt",
  price: 2314,
  ratings: 334,
  stars: 2,
  likes: 380,
  category: "Men's Clothing",
  styles: [
    {
      description: "black",
      picture: "/images/poloshirtblack.webp",
      isCurrentStyle: true,
    },
    {
      description: "blue",
      picture: "/images/bluepoloshirt.jpg",
      isCurrentStyle: false,
    },
  ],
};
//   const menClothingProducts = [products[0], products[2]];
const emptyCart: Cart = [];
const productsURL = "http://localhost:3000/products";
const cartURL = "http://localhost:3000/cart";
const menClothingURL = "http://localhost:3000/products?category=Men's Clothing";
const shoppingProductRoute: RouteObject = {
  path: "/product/:id",
  element: <ShoppingProduct />,
};

beforeAll(() => {
  HTMLElement.prototype.scrollIntoView = vi.fn();

  vi.spyOn(global, "fetch").mockImplementation(
    vi.fn((url: string) => {
      const values =
        url === productsURL
          ? product
          : cartURL === url
          ? emptyCart
          : // : url === menClothingURL
            // ? menClothingProducts
            null;
      if (!values) throw new Error("Couldn't process url string");
      return Promise.resolve({
        json: () => Promise.resolve(values),
      });
    }) as Mock
  );

  const mock = vi.hoisted(() => ({
    useLocation: vi.fn(() => ({
      state: product,
    })),
  }));

  vi.mock("react-router-dom", async () => {
    const routerData = await vi.importActual("react-router-dom");

    return {
      ...routerData,
      useLocation: mock.useLocation,
    };
  });
});

afterAll(() => {
  vi.resetAllMocks();
});

describe("ShoppingProduct", () => {
  it("renders the product's image", async () => {


    // const router = createMemoryRouter([shoppingProductRoute]);
    render(<ShoppingProduct />, { wrapper: BrowserRouter });

    const img = (await waitFor(() =>
      screen.getByRole("img")
    )) as HTMLImageElement;

    expect(img.src).toBe("http://localhost:3000/images/poloshirtblack.webp");
    expect(img).toMatchSnapshot();
  });
});
