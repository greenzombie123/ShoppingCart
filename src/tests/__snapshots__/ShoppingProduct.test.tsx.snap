// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShoppingProduct > renders the product's image 1`] = `
<img
  alt="polo shirt black"
  src="/images/poloshirtblack.webp"
/>
`;
